Adds more context to build and update docs on home page (#31674)
expo/expo: docs/pages/deploy/build-project.mdx

description: Learn how to create a production build for your app that is ready to be submitted to app stores from the command line using EAS Build.
sidebar_title: Build project for app stores
---

import { BookOpen02Icon } from '@expo/styleguide-icons/outline/BookOpen02Icon';

import { BoxLink } from '~/ui/components/BoxLink';
import { Collapsible } from '~/ui/components/Collapsible';
import { Terminal } from '~/ui/components/Snippet';
import { Tabs, Tab } from '~/ui/components/Tabs';

Whether you have built a native app binary using [EAS](/build/setup/) or [locally](/guides/local-app-development/), the next step in your app development journey is to submit your app to the stores. To do so, you need to create a **production build**.

Production builds are submitted to app stores for release to the general public or as part of a store-facilitated testing process such as TestFlight. This guide explains how to create production builds with [EAS](#production-builds-using-eas) and [locally](#production-builds-locally). It is also possible to create production builds for Expo apps with any CI service capable of compiling Android and iOS apps.

## Production builds using EAS

> Production builds must be installed through their respective app stores. They cannot be installed directly on your Android Emulator or device, or iOS Simulator or device. The only exception to this is if you explicitly set `"buildType": "apk"` for Android on your build profile. However, it is recommended to use **aab** when submitting to stores, and this is the default configuration.

Reviewer: ```suggestion
Production builds must be installed through their respective app stores. You cannot install them directly on your Android Emulator, iOS Emulator, or device. The only exception to this is if you explicitly set `"buildType": "apk"` for Android on your build profile. However, it is recommended to use **aab** when submitting to stores, and this is the default configuration.
```

- In the second sentence, we can avoid using unnecessary passive voice usage and repetitiveness.

### `eas.json` configuration

> A minimal configuration for building a production build in **eas.json** that is already created when you create your first build:

Reviewer: ```suggestion
A minimal configuration for building a production build in **eas.json** is already created when you create your first build:
```

Unnecessary since it breaks the flow of the text.

{/* prettier-ignore */}
```json eas.json
{
  "build": {
    /* @hide ... */ /* @end */
    "production": {}
    /* @hide ... */ /* @end */
  }
}
```

### Create a production build

To create a production build, run the following command for a platform:

<Tabs tabs={['Android', 'iOS']}>
  <Tab>
    <Terminal cmd={['$ eas build --platform android']} />
  </Tab>
  <Tab>
    <Terminal cmd={['$ eas build --platform ios']} />
  </Tab>
</Tabs>

> > You can attach a message to the build by passing `--message` to the build command, for example, `eas build --platform ios --message "Some message"`. The message will appear on the website. It comes in handy when you want to leave a note with the purpose of the build for your team.

Reviewer: ```suggestion
> You can attach a message to the build by passing `--message` to the build command, for example, `eas build --platform ios --message "Some message"`. The message will appear on the Expo dashboard. It comes in handy when you want to specify the purpose of the build for your team.
```

- In the above, the "website" is vague. Let's replace it with the Expo dashboard.
- Let's remove the wordiness in the last sentence.

> > You can attach a message to the build by passing `--message` to the build command, for example, `eas build --platform ios --message "Some message"`. The message will appear on the website. It comes in handy when you want to leave a note with the purpose of the build for your team.

Reviewer: I don't think this should be a callout. It should be part of the main content. Also, specifying the command outside the callout breaks the flow as I am scanning the content.

> Alternatively, you can use `--platform all` option to build for Android and iOS at the same time:

<Terminal cmd={['$ eas build --platform all']} />

## Developer account

You will need to have a developer account for the app store you want to submit your app.

<Collapsible summary="Google Play Developer membership is required to distribute to the Google Play Store.">

You can build and sign your app using EAS Build, but you can't upload it to the Google Play Store unless you have a membership, a one-time $25 USD fee.

</Collapsible>

<Collapsible summary="Apple Developer Program membership is required to build for the Apple App Store.">

> If you are going to use EAS Build to create release builds for the Apple App Store, you need access to an account with a $99 USD [Apple Developer Program](https://developer.apple.com/programs) membership.

Reviewer: ```suggestion
If you are going to use EAS Build to create release builds for the Apple App Store, you need access to an account with a $99 USD [Apple Developer Program](https://developer.apple.com/programs) membership.
```

- We are using the term "release" builds here but in this doc we don't specify that production builds are aka release builds. Maybe we should do it in the introduction.

Author: I changed this to be "production" instead of "release"

</Collapsible>

## App signing credentials

> Before the build process can start for app stores, you will need to have a store developer account and generate or provide app signing credentials.

Reviewer: ```suggestion
Before the build process can start for app stores, you need a store developer account and generate or provide app signing credentials.
```

> Whether you have experience with generating app signing credentials or not, EAS CLI does the heavy lifting. You can opt-in for EAS CLI to handle the app signing credentials process.

Reviewer: ```suggestion
Whether you have experience with generating app signing credentials or not, EAS CLI can do the heavy lifting. You can opt-in for EAS CLI to handle the app signing credentials process.
```

As a reader, "does the heavy lifting" sounds like I should be aware of EAS CLI features instead of being told what to use.

### Android app signing credentials

> - If you have not yet generated a keystore for your app, you can let EAS CLI take care of that for you by selecting `Generate new keystore`, and then you are done. The keystore is stored securely on EAS servers.

Reviewer: ```suggestion
- If you have not yet generated a keystore for your app, use EAS CLI by selecting `Generate new keystore`, and then you are done. The keystore is stored securely on EAS servers.
```

Let's be more direct in the first sentence.

- If you want to manually generate your keystore, see the [manual Android credentials guide](/app-signing/local-credentials#android-credentials) for more information.

### iOS app signing credentials

> - If you have not generated a provisioning profile and/or distribution certificate yet, you can let EAS CLI take care of that for you by signing in to your Apple Developer Program account and following the prompts.

Reviewer: ```suggestion
- If you have not generated a provisioning profile and/or distribution certificate yet, use EAS CLI by signing in to your Apple Developer Program account and following the prompts.
```

Same as above.

- If you want to manually generate your credentials, see the [manual iOS credentials guide](/app-signing/local-credentials#ios-credentials) for more information.

## Wait for the build to complete

> By default, the `eas build` command will wait for your build to complete, but you can interrupt it if you prefer not to wait. Monitor the progress and read the logs by following the link to the build details page that EAS CLI prompts once the build process gets started. You can also find this page by visiting [your build dashboard](https://expo.dev/builds) or running the following command:

Reviewer: ```suggestion
By default, the `eas build` command will wait for your build to complete, but you can interrupt it if you prefer not to wait. Instead, use the builds details page link prompted by EAS CLI to monitor the build progress and read the build logs. You can also find this page by visiting [your build dashboard](https://expo.dev/builds) or running the following command:
```

Let's improve the readability of the second sentence. There's no need to mention "build process gets started". We are already doing that in the first sentence.

<Terminal cmd={['$ eas build:list']} />

If you are a member of an organization and your build is on its behalf, you will find the build details on [the build dashboard for that account](https://expo.dev/accounts/[account]/builds).

## Production builds locally

> To create a production build locally, see the following React Native guides for more information on the necessary steps that are required for Android and iOS.

Reviewer: ```suggestion
To create a production build locally, see the following React Native guides for more information on the necessary steps for Android and iOS.
```

Remove wordiness.
